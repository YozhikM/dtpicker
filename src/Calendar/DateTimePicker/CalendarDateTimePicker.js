/* @flow */

import React from 'react';
import SvgIcon from '../../SvgIcon/SvgIcon';
import CalendarDate from '../DateTimePicker/CalendarDate';
import DateMaskInput from '../DateMaskInput';
import s from './Calendar.scss';

type Props = {
  value: Date,
  visibleDate?: Date,
  highlight: Array<Date> | Date,
  onChange?: Date => void,
  onSetDate?: Date => void,
  onSetDateByClick?: Date => void,
  time?: boolean,
  icon?: boolean,
  leftArrow?: boolean,
  rightArrow?: boolean,
  isCalendarShown?: boolean,
};

type State = {
  highlight?: Array<Date> | Date,
  value: Date,
  visibleDate: Date,
  isCalendarShown: boolean,
};

class CalendarDateTimePicker extends React.Component<Props, State> {
  constructor(props: Props) {
    super(props);

    this.state = {
      visibleDate: this.props.visibleDate || new Date(),
      value: this.props.value,
      isCalendarShown: !!this.props.isCalendarShown,
    };
  }

  componentWillReceiveProps(nextProps: Props) {
    if (nextProps.visibleDate !== this.state.visibleDate)
      this.setState({ visibleDate: nextProps.visibleDate });

    if (nextProps.value !== this.state.value) {
      this.setState({ value: nextProps.value });
    }

    if (nextProps.isCalendarShown !== this.state.isCalendarShown) {
      this.setState({ isCalendarShown: nextProps.isCalendarShown });
    }
  }

  onChangeInputValue = (date: Date) => {
    const { onChange, onSetDate } = this.props;

    if (date && onSetDate) {
      onSetDate(date);
    }
    if (date && onChange) {
      onChange(date);
    }
  };

  onChange = (value: Date) => {
    const { onChange } = this.props;

    if (onChange) {
      onChange(value);
    }
  };

  onSetDate = (date: Date) => {
    const { onSetDateByClick } = this.props;

    if (onSetDateByClick) onSetDateByClick(date);
  };

  onSetTime = (time: Date) => {
    this.setState({
      visibleDate: time,
    });
  };

  render() {
    const {
      highlight,
      icon,
      leftArrow,
      rightArrow,
      time,
      isCalendarShown,
      visibleDate,
    } = this.props;
    const { value } = this.state;
    const style = {
      width: '80%',
    };

    return (
      <div className={s.container}>
        <DateMaskInput style={style} visibleDate={visibleDate} onChange={this.onChangeInputValue} />
        {icon && (
          <div className={s.icon}>
            <SvgIcon file="calendar" />
          </div>
        )}
        {isCalendarShown && (
          <CalendarDate
            highlight={highlight}
            value={value}
            visibleTime={visibleDate}
            onSetDate={this.onSetDate}
            onSetTime={this.onSetTime}
            onChange={this.onChange}
            leftArrow={leftArrow}
            rightArrow={rightArrow}
            time={time}
          />
        )}
      </div>
    );
  }
}

export default CalendarDateTimePicker;
