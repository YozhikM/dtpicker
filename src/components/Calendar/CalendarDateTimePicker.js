/* @flow */

import React from 'react';
import CalendarDate from '../Calendar/CalendarDate';
import MaskedInput from 'react-text-mask';
import SvgIcon from '../SvgIcon';
import s from './Calendar.scss';
import { format, isValid, isSameDay } from 'date-fns';

type Props = {|
  highlight: Array<Date> | Date,
  value: Date,
  visibleDate?: Date,
  time?: boolean,
  onChangeCalendarVisibility?: boolean => void,
  isCalendarShown?: boolean,
  onChange?: (value: Date) => void,
  borderLeft?: boolean,
  borderRight?: boolean,
  icon?: boolean,
  onSetDate: (date: Date) => void,
  leftArrow?: boolean,
  rightArrow?: boolean,
|};
type State = {|
  highlight: Array<Date> | Date,
  value: Date,
  visibleDate: Date,
  isCalendarShown: boolean,
  inputValue: string,
|};

class CalendarDateTimePicker extends React.Component<Props, State> {
  constructor(props: Props) {
    super(props);

    this.state = {
      highlight: this.props.highlight,
      visibleDate: this.props.visibleDate,
      value: this.props.value,
      isCalendarShown: !!this.props.isCalendarShown,
      inputValue: '',
    };
  }

  formatWithTime = (date: Date) => {
    return format(date, 'DD/MM/YYYY HH:mm');
  };
  formatWithoutTime = (date: Date) => {
    return format(date, 'DD/MM/YYYY');
  };

  componentWillMount() {
    const { visibleDate, time } = this.props;
    if (time) {
      this.setState({ inputValue: this.formatWithTime(visibleDate) });
    } else {
      this.setState({
        inputValue: this.formatWithoutTime(visibleDate),
      });
    }
  }

  componentWillReceiveProps(nextProps: Props) {
    const { time } = this.props;
    if (nextProps.time !== time) {
      if (time) {
        this.setState({
          inputValue: this.formatWithTime(nextProps.visibleDate),
        });
      } else {
        this.setState({
          inputValue: this.formatWithoutTime(nextProps.visibleDate),
        });
      }
    }
    if (
      nextProps.visibleDate !== this.state.visibleDate &&
      !isSameDay(nextProps.visibleDate, this.props.value)
    ) 
		this.setState({ visibleDate: nextProps.visibleDate });
	{
      if (time) {
        this.setState({
          inputValue: this.formatWithTime(nextProps.visibleDate),
        });
      } else {
        this.setState({
          inputValue: this.formatWithoutTime(nextProps.visibleDate),
        });
      }
    }

    if (nextProps.value !== this.state.value) {
      this.setState({ value: nextProps.value });
    }

    if (nextProps.isCalendarShown !== this.state.isCalendarShown) {
      this.setState({ isCalendarShown: nextProps.isCalendarShown });
    }
  }

  onChangeInputValue = (e: any) => {
    const { onChange, onSetDate, time } = this.props;
    const value = e.target.value;
    this.setState({ inputValue: value });
    let userDate;
    const YY = Number(value.substr(6, 4));
    const MM = Number(value.substr(3, 2)) - 1;
    const DD = Number(value.substr(0, 2));
    const HH = Number(value.substr(11, 2));
    const mm = Number(value.substr(14, 2));
    if (time) {
      userDate = new Date(YY, MM, DD, HH, mm);
    } else {
      userDate = new Date(YY, MM, DD);
    }
    if (value.length > 0 && onSetDate && isValid(userDate)) {
      onSetDate(userDate);
      onChange(userDate);
    }
  };

  onFocusInput = () => {
    const { isCalendarShown } = this.state;
    if (!isCalendarShown) {
      this.onChangeVisibility();
    }
  };

  onPressEnter = (e: Event) => {
    if (e.key === 'Enter') {
      this.onChangeVisibility();
    }
  };

  onChangeVisibility = () => {
    const { isCalendarShown } = this.state;
    const { onChangeCalendarVisibility } = this.props;
    if (onChangeCalendarVisibility) {
      onChangeCalendarVisibility(!isCalendarShown);
    }
  };

  onChange = (value: Date) => {
    const { onChange } = this.props;
    if (onChange) {
      onChange(value);
    }
  };

  onSetDate = (date: Date) => {
    const { time } = this.props;
    if (time) {
      this.setState({
        inputValue: this.formatWithTime(date),
      });
    } else {
      this.setState({
        inputValue: this.formatWithoutTime(date),
      });
    }
    const { onSetDate } = this.props;
    if (onSetDate) onSetDate(date);
  };

  onSetTime = (time: Date) => {
    this.setState({
      visibleDate: time,
      inputValue: this.formatWithTime(time),
    });
  };

  render() {
    const {
      highlight,
      borderLeft,
      borderRight,
      icon,
      leftArrow,
      rightArrow,
      time,
      isCalendarShown,
      visibleDate,
	  isSingleCalendar
    } = this.props;
    const { inputValue, value } = this.state;
    const iconStyle = {
      display: icon ? 'flex' : 'none',
    };
    const borderStyle = {
      borderLeft: borderLeft ? '3px solid #34495e' : 0,
      borderRight: borderRight ? '3px solid #34495e' : 0,
    };
    let mask;
    if (time) {
      mask = [
        /[0-9]/,
        /\d/,
        '/',
        /\d/,
        /\d/,
        '/',
        /\d/,
        /\d/,
        /\d/,
        /\d/,
        ' ',
        /\d/,
        /\d/,
        ':',
        /\d/,
        /\d/,
      ];
    } else {
      mask = [
        /[0-9]/,
        /\d/,
        '/',
        /\d/,
        /\d/,
        '/',
        /\d/,
        /\d/,
        /\d/,
        /\d/,
      ];
    }

    return (
      <div className={s.container}>
        <MaskedInput
          mask={mask}
          className={s.input}
          style={borderStyle}
          type="text"
          onChange={this.onChangeInputValue}
          onFocus={this.onFocusInput}
          onKeyPress={this.onPressEnter}
          value={inputValue}
          showMask
        />
        <div
          onClick={this.onChangeVisibility}
          className={s.icon}
          style={iconStyle}
        >
          <SvgIcon file="calendar" />
        </div>
        {isCalendarShown && (
          <CalendarDate
            highlight={highlight}
            value={value}
            visibleTime={visibleDate}
            onSetDate={this.onSetDate}
            onSetTime={this.onSetTime}
            onChange={this.onChange}
            leftArrow={leftArrow}
            rightArrow={rightArrow}
            time={time}
			isSingleCalendar={isSingleCalendar}
          />
        )}
      </div>
    );
  }
}

export default CalendarDateTimePicker;
