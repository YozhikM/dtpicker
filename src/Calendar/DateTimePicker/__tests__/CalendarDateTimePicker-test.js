/* @flow */

import React from 'react';
import CalendarDateTimePicker from '../CalendarDateTimePicker';

describe('CalendarDateTimePicker', () => {
  describe('render()', () => {
    it('basic', () => {
      const date = new Date(2017, 9, 8);
      expect(
        shallow(
          <CalendarDateTimePicker highlight={date} value={new Date(2001, 3)} isCalendarShown />
        )
      ).toMatchSnapshot();
    });

    it('if time', () => {
      const date = new Date(2017, 9, 8);
      expect(
        shallow(
          <CalendarDateTimePicker highlight={date} value={new Date(2001, 3)} isCalendarShown time />
        )
      ).toMatchSnapshot();
    });

    it('if !isCalendarShown', () => {
      expect(
        shallow(
          <CalendarDateTimePicker highlight={new Date(2017, 9, 8)} value={new Date(2001, 3)} />
        )
      ).toMatchSnapshot();
    });

    it('if icon', () => {
      expect(
        shallow(
          <CalendarDateTimePicker
            highlight={new Date(2017, 9, 8)}
            value={new Date(2001, 3)}
            isCalendarShown
            icon
          />
        )
      ).toMatchSnapshot();
    });

    it('if leftArrow', () => {
      expect(
        shallow(
          <CalendarDateTimePicker
            highlight={new Date(2017, 9, 8)}
            value={new Date(2001, 3)}
            isCalendarShown
            leftArrow
          />
        )
      ).toMatchSnapshot();
    });

    it('if rightArrow', () => {
      expect(
        shallow(
          <CalendarDateTimePicker
            highlight={new Date(2017, 9, 8)}
            value={new Date(2001, 3)}
            isCalendarShown
            rightArrow
          />
        )
      ).toMatchSnapshot();
    });
  });

  describe('componentWillReceiveProps()', () => {
    it('visibleDate', () => {
      const wrapper = shallow(
        <CalendarDateTimePicker highlight={new Date(2010, 1, 11)} value={new Date(2011, 9, 7)} />
      );
      const nextProps = {
        visibleDate: new Date(2010, 0, 10),
        highlight: new Date(),
        value: new Date(),
      };
      wrapper.instance().componentWillReceiveProps(nextProps);

      expect(wrapper.state('visibleDate')).toEqual(new Date('2010-01-09T18:00:00.000Z'));
    });

    it('value', () => {
      const wrapper = shallow(
        <CalendarDateTimePicker
          highlight={new Date(2060, 5, 11)}
          visibleDate={new Date(2060, 5, 11)}
          value={new Date(2051, 3, 1)}
          onSetDate={jest.fn()}
        />
      );
      const nextProps = {
        highlight: new Date(),
        value: new Date(2010, 0, 10),
      };
      wrapper.instance().componentWillReceiveProps(nextProps);

      expect(wrapper.state('value')).toEqual(new Date(2010, 0, 10));
    });

    it('isCalendarShown', () => {
      const wrapper = shallow(
        <CalendarDateTimePicker
          highlight={new Date(2060, 5, 11)}
          value={new Date(2051, 3, 1)}
          isCalendarShown={false}
        />
      );
      const nextProps = {
        highlight: new Date(),
        value: new Date(2010, 0, 10),
        isCalendarShown: true,
      };
      wrapper.instance().componentWillReceiveProps(nextProps);

      expect(wrapper.state('isCalendarShown')).toBe(true);
    });
  });

  it('onChange()', () => {
    const onChange = jest.fn();
    const wrapper = shallow(
      <CalendarDateTimePicker
        highlight={new Date()}
        value={new Date()}
        onChange={onChange}
        isCalendarShown
      />
    );
    const newDate = new Date(2019, 8, 3);
    wrapper
      .find('CalendarDate')
      .props()
      .onChange(newDate);

    expect(onChange).toBeCalledWith(newDate);
  });

  it('onSetDate()', () => {
    const onSetDateByClick = jest.fn();
    const wrapper = shallow(
      <CalendarDateTimePicker
        highlight={new Date()}
        visibleDate={new Date()}
        value={new Date()}
        onSetDateByClick={onSetDateByClick}
        isCalendarShown
      />
    );
    const newDate = new Date(2009, 2, 3);
    wrapper
      .find('CalendarDate')
      .props()
      .onSetDate(newDate);

    expect(onSetDateByClick).toBeCalledWith(newDate);
  });

  it('onSetTime()', () => {
    const wrapper = shallow(
      <CalendarDateTimePicker highlight={new Date()} value={new Date()} isCalendarShown time />
    );
    const newDate = new Date('2019-12-12T18:00:00.000Z');
    wrapper
      .find('CalendarDate')
      .props()
      .onSetTime(newDate);

    expect(wrapper.state('visibleDate')).toEqual(newDate);
  });

  it('onChangeInputValue()', () => {
    const onSetDate = jest.fn();
    const onChange = jest.fn();
    const wrapper = shallow(
      <CalendarDateTimePicker
        highlight={new Date(2009, 1, 10)}
        value={new Date(2111, 5, 3)}
        onSetDate={onSetDate}
        onChange={onChange}
      />
    );
    const newDate = new Date('2019-12-09T18:00:00.000Z');
    wrapper
      .find('DateMaskInput')
      .props()
      .onChange(newDate);

    expect(onSetDate).toBeCalledWith(newDate);
    expect(onChange).toBeCalledWith(newDate);
  });
});
