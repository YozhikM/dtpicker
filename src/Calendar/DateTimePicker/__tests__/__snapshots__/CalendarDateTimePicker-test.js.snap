// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CalendarDateTimePicker render() basic 1`] = `
<div>
  <DateMaskInput
    onChange={[Function]}
    style={
      Object {
        "width": "80%",
      }
    }
  />
  <CalendarDate
    highlight={2017-10-07T18:00:00.000Z}
    onChange={[Function]}
    onSetDate={[Function]}
    onSetTime={[Function]}
    value={2001-03-31T17:00:00.000Z}
  />
</div>
`;

exports[`CalendarDateTimePicker render() if !isCalendarShown 1`] = `
<div>
  <DateMaskInput
    onChange={[Function]}
    style={
      Object {
        "width": "80%",
      }
    }
  />
</div>
`;

exports[`CalendarDateTimePicker render() if icon 1`] = `
<div>
  <DateMaskInput
    onChange={[Function]}
    style={
      Object {
        "width": "80%",
      }
    }
  />
  <div>
    <SvgIcon
      file="calendar"
    />
  </div>
  <CalendarDate
    highlight={2017-10-07T18:00:00.000Z}
    onChange={[Function]}
    onSetDate={[Function]}
    onSetTime={[Function]}
    value={2001-03-31T17:00:00.000Z}
  />
</div>
`;

exports[`CalendarDateTimePicker render() if leftArrow 1`] = `
<div>
  <DateMaskInput
    onChange={[Function]}
    style={
      Object {
        "width": "80%",
      }
    }
  />
  <CalendarDate
    highlight={2017-10-07T18:00:00.000Z}
    leftArrow={true}
    onChange={[Function]}
    onSetDate={[Function]}
    onSetTime={[Function]}
    value={2001-03-31T17:00:00.000Z}
  />
</div>
`;

exports[`CalendarDateTimePicker render() if rightArrow 1`] = `
<div>
  <DateMaskInput
    onChange={[Function]}
    style={
      Object {
        "width": "80%",
      }
    }
  />
  <CalendarDate
    highlight={2017-10-07T18:00:00.000Z}
    onChange={[Function]}
    onSetDate={[Function]}
    onSetTime={[Function]}
    rightArrow={true}
    value={2001-03-31T17:00:00.000Z}
  />
</div>
`;

exports[`CalendarDateTimePicker render() if time 1`] = `
<div>
  <DateMaskInput
    onChange={[Function]}
    style={
      Object {
        "width": "80%",
      }
    }
  />
  <CalendarDate
    highlight={2017-10-07T18:00:00.000Z}
    onChange={[Function]}
    onSetDate={[Function]}
    onSetTime={[Function]}
    time={true}
    value={2001-03-31T17:00:00.000Z}
  />
</div>
`;
